### PR TITLE
Schema - Fix boolean fields in civicrm_event and civicrm_contribution

### PR DESCRIPTION
**Pull request by colemanw**
Overview
----------------------------------------
This fixes 2 tables per @eileenmcnaughton's [comment here](https://github.com/civicrm/civicrm-core/pull/22430#issuecomment-1008432906): "boolean fields should all be required & have a default."

Before
----------------------------------------
When the `<required>` flag is not set, boolean fields are actually trinary, as they can be `1`, `0`, or `NULL`. This can cause bugs, and slow queries that work around the bugs using e.g. `( field = 0 OR field IS NULL )`.

After
----------------------------------------
At least for these two tables, all boolean fields now consistently have a default and cannot be NULL.

Comments
----------------------------------------
This was a PITA to do manually. It took about an hour of my time and the end result is only 2 tables have been fixed, out of... a lot more than 2! I'm frustrated that this bad pattern wasn't noticed until now and has been copied & pasted all over our schema!
I really wish we were using something like doctrine which would auto-generate these `ALTER TABLE` upgrade statements for us whenever we change the schema. That would make this process a lot less tedious.

Frustration aside, I think we really need to fix this for more than just these 2 tables (this PR serves as an example for further cleanup). Because we want to avoid bugs and slow queries in SearchKit, etc.

One concrete thing I think would be a great idea: when our sql generator writes the `CREATE TABLE` statement, when parsing the xml it should supply defaults of `required: true, default: 0` for fields of `type: boolean`. That would update every boolean field in our schema in one shot, and make it impossible for future devs to keep making this mistake. However, we'd have to consider the implications of such a move: it would require upgrade sql for all the affected fields, and we'd also have to put the word out to extension devs that they should re-run `civix generate:entity-boilerplate` and write any necessary upgrade to fix the boolean fields in their extensions' schema.

### CODE REVIEW

File: CRM/Core/BAO/CustomGroup.php [hunk 1]
```diff
@@ -2259,7 +2259,7 @@ public static function getSubTypes(): array {
     $sel2['Campaign'] = $campaignTypes;
     $sel2['Membership'] = $membershipType;
     $sel2['ParticipantRole'] = $participantRole;
-    $sel2['ParticipantEventName'] = CRM_Event_PseudoConstant::event(NULL, FALSE, "( is_template IS NULL OR is_template != 1 )");
+    $sel2['ParticipantEventName'] = CRM_Event_PseudoConstant::event(NULL, FALSE, "( is_template != 1 )");
```
Comment (colemanw) [inline CRM/Core/BAO/CustomGroup.php:2262]: This. This right here is the reason we need to fix our schema. No more ugly workarounds for unreliable boolean fields!

**Comment by colemanw** [comment]
@monishdeb can you please review this PR, and then can you think of a way to abstract this and apply this fix to our entire schema? Here's an idea:

1. After merging this PR, create a clean install on `master`.
2. Make a backup copy of `sql/civicrm.mysql`.
3. Make the change to GenCode I suggested above (when parsing the xml, supply defaults for fields of type `boolean`).
4. Run `setup.sh -g` to regenerate DAO files (new defaults should get reflected in those files with `required` and `default` added for booleans, even without any updates to the schema xml files). That also ought to regenerate `sql/civicrm.mysql`.
5. Use a diff tool between your backup and the new `sql/civicrm.mysql`.
6. Grab the changed columns from the diff and use some regex magic to convert the `CREATE TABLE` statements into `ALTER TABLE` statements to be placed in `5.49.alpha1.mysql.tpl`.

File: CRM/Core/BAO/CustomGroup.php (hunk 1 above)
Comment (demeritcowboy) [reply CRM/Core/BAO/CustomGroup.php:2262]: I wonder if there's an argument to treat these kind of schema changes the same as deprecations, where there's a period of time before actually changing the code that uses the old value. Like `// @deprecated can remove check for null in 5.55`.

I don't have a specific instance for this code line, but there's been issues before where upgrade code in an intermediate version between the initial and final version will call some core code such as this indirectly, but the upgrade step that this depends on hasn't run yet. E.g.
* Upgrading from 5.47 to 5.49
* Pretend that 5.48 had some upgrade code that hits this line.
* At the point that upgrade code runs, is_template could still be null in the db, but the line has the new code.
* Here it won't crash, but might not work as intended.

Comment (colemanw) [reply CRM/Core/BAO/CustomGroup.php:2262]: @demeritcowboy I think that's not a bad idea, but this particular field has already been fixed once (just not completely - the upgrade script forgot to add `NOT NULL`) and this code only runs on the form AFAIK.

https://github.com/civicrm/civicrm-core/commit/99f13d8d83f1786a6bd8e2238d0424663488a988#diff-77e9eaeaf4c5c2129efa23ff414bf1657535b3f6e50e66324457f17095c666a6R254-R256

**Comment by monishdeb** [comment]
* General standards
    * ([`r-explain`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-explain))  __PASS__ (I can agree with the assessment and first noticed this issue from a slow query https://github.com/civicrm/civicrm-core/pull/3005 )
    * ([`r-user`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-user)) PASS
    * ([`r-doc`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-doc)) PASS
    * ([`r-run`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-run)) __PASS__ Tested on local
* Developer standards
    * ([`r-tech`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-tech)) __PASS__ 
    * ([`r-code`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-code)) __PASS__
    * ([`r-maint`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-maint)) __PASS__
    * ([`r-test`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-test)) __PASS__

**Comment by monishdeb** [comment]
I will submit a followup PR to replicate this necessary fix for other boolean fields. Thanks for the detail step @colemanw  :)

**Comment by magnolia61** [comment]
Are custom required boolean fields also in scope?

**Comment by colemanw** [comment]
> Thanks for the detail step @colemanw :)

I hope that idea works @monishdeb. If not, maybe something similar :)

**Comment by colemanw** [comment]
@magnolia61 that does look like a related issue. Maybe that could be another followup @monishdeb - it looks like the column "not null" is being suppressed here, where it actually should be forced TRUE for boolean fields (and they should be forced to have a default)

https://github.com/civicrm/civicrm-core/blob/f4ba925d60b8fb359faed7288067e4090b350f75/CRM/Core/BAO/CustomField.php#L1812-L1815

**Pull request merged**